### DISCUSSION TITLE
Looking for plugins showing opened windows in the space

### DISCUSSION THREAD

**Discussion by tahsinrahman** [FelixKratz/SketchyBar#280] [Q&A] [answered]
Is there's any plugin that dynamically shows opened app/windows in current space? 
Similar to output of this command `yabai -m query --windows --space | jq .[].title`.

I went through the plugin list discussion, but couldn't find any.

**Comment by FelixKratz** [comment, marked as answer]
I have something like this in my dotfiles. Basically, what you need is this:

- A script that updates the apps on spaces and gives them an icon (here it is done in `windows_on_spaces` function with an event that has the same name): https://github.com/FelixKratz/dotfiles/blob/master/.config/sketchybar/plugins/yabai.sh
- An icon mapping from application name to an actual icon (in this case using [sketchybar-app-font](https://github.com/kvndrsslr/sketchybar-app-font)): https://github.com/FelixKratz/dotfiles/blob/master/.config/sketchybar/plugins/icon_map.sh
- You need to trigger the update event when an app is created/destroyed/moved etc. e.g.
```bash
yabai -m signal --add event=window_created action="sketchybar --trigger windows_on_spaces"
yabai -m signal --add event=window_destroyed action="sketchybar --trigger windows_on_spaces"
```
and in skhd when moving windows to different spaces, e.g. (search for `windows_on_spaces`):
https://github.com/FelixKratz/dotfiles/blob/master/.config/skhd/skhdrc

Now you can decide when to show those icons, which are contained in the space items label.

**Reply by tahsinrahman** [reply]
Hi @FelixKratz , thanks for the reply! Though I was looking for app names/icons in the title bar instead of spaces section, I think its a good starting pointer and I hope i can find my way from here.

**Reply by FelixKratz** [reply]
> Hi @FelixKratz , thanks for the reply! Though I was looking for app names/icons in the title bar instead of spaces section, I think its a good starting pointer and I hope i can find my way from here.

Ah ok, I think with everything linked here you will be able to make what you are imagining work. When you are done, don't hesitate to show the finished setup, it surely is a novel idea.

**Reply by tahsinrahman** [reply]
I could get it work! 

But only one nitpick, when i switch to an empty space where there's no app opened, it shows the titles from previous space.

here're my configs, did i miss anything?

sketchybarrc
```bash
sketchybar -m --add item title left \
              --set title script="$HOME/.config/sketchybar/plugins/window_title.py" \
              --subscribe title space_change front_app_switched
```

plugins/window_title.py
```python
#!/usr/bin/env python3

import subprocess
import json

all_apps_cmd = '''
yabai -m query --windows --space | jq '.[] | {app: .app, pid: .pid, index: ."stack-index"}' | jq -s 'sort_by(.index)'
'''
apps = subprocess.run(all_apps_cmd, capture_output=True, shell=True, text=True).stdout
apps = json.loads(apps)

current_pid_cmd = '''
yabai -m query --windows --window | jq '.pid'
'''
current_pid = subprocess.run(current_pid_cmd, capture_output=True, shell=True, text=True)
current_pid = int(current_pid.stdout)

display = []
for (idx, app) in enumerate(apps):
    app_name = app["app"]
    if app["pid"] == current_pid:
        app_name = '*' + app_name
    display.append(app_name)

labels = ' | '.join(display)
subprocess.run('sketchybar -m --set title label="{}"'.format(labels), shell=True)
```

---

i also tried custom events, and get them triggered from yabai when space is changed, but still doesn't work

sketchybarrc
```bash
sketchybar -m --add event title_change

sketchybar -m --add item title left \
              --set title script="$HOME/.config/sketchybar/plugins/window_title.py" \
              --subscribe title title_change space_change front_app_switched
```

yabairc
```bash
yabai -m signal --add event=space_changed action="sketchybar -m --trigger title_change &> /dev/null"
```
